client/src/rr/dnssec/digest_type.rs

```diff
@@ -62,30 +66,78 @@ impl DigestType {
 
     /// The OpenSSL counterpart for the digest
     #[cfg(feature = "openssl")]
-    pub fn to_openssl_digest(&self) -> DnsSecResult<MessageDigest> {
+    pub fn to_openssl_digest(&self) -> DnsSecResult<hash::MessageDigest> {
         match *self {
-            DigestType::SHA1 => Ok(MessageDigest::sha1()),
-            DigestType::SHA256 => Ok(MessageDigest::sha256()),
-            DigestType::SHA384 => Ok(MessageDigest::sha384()),
-            DigestType::SHA512 => Ok(MessageDigest::sha512()),
+            DigestType::SHA1 => Ok(hash::MessageDigest::sha1()),
+            DigestType::SHA256 => Ok(hash::MessageDigest::sha256()),
+            DigestType::SHA384 => Ok(hash::MessageDigest::sha384()),
+            DigestType::SHA512 => Ok(hash::MessageDigest::sha512()),
             _ => {
                 Err(DnsSecErrorKind::Msg(format!("digest not supported by openssl: {:?}", self))
                         .into())
             }
         }
     }
 
+    /// The *ring* counterpart for the digest
+    #[cfg(feature = "ring")]
+    pub fn to_ring_digest_alg(&self) -> DnsSecResult<&'static digest::Algorithm> {
+        match *self {
+            DigestType::SHA1 => Ok(&digest::SHA1),
+            DigestType::SHA256 => Ok(&digest::SHA256),
+            DigestType::SHA384 => Ok(&digest::SHA384),
+            DigestType::SHA512 => Ok(&digest::SHA512),
+            _ =>
+                Err(DnsSecErrorKind::Msg(format!("digest not supported by ring: {:?}", self))
+                    .into())
+        }
+    }
+
     /// Hash the data
     #[cfg(feature = "openssl")]
-    pub fn hash(&self, data: &[u8]) -> DnsSecResult<DigestBytes> {
+    pub fn hash(&self, data: &[u8]) -> DnsSecResult<Digest> {
         hash::hash2(try!(self.to_openssl_digest()), data).map_err(|e| e.into())
     }
 
+    /// Hash the data
+    #[cfg(feature = "ring")]
+    pub fn hash(&self, data: &[u8]) -> DnsSecResult<Digest> {
+        let alg = try!(self.to_ring_digest_alg());
+        Ok(digest::digest(alg, data))
+    }
+
     /// This will always error, enable openssl feature at compile time
-    #[cfg(not(feature = "openssl"))]
+    #[cfg(not(any(feature = "openssl", feature = "ring")))]
     pub fn hash(&self, _: &[u8]) -> DnsSecResult<Vec<u8>> {
         Err(DnsSecErrorKind::Message("openssl feature not enabled").into())
```

Comment:
maybe change this error to say "openssl or *ring*"?

Reply:
Done.